File: Python/previous-permutation-with-one-swap.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def prevPermOpt1(self, A):
        """
        :type A: List[int]
        :rtype: List[int]
        """
        for left in reversed(xrange(len(A)-1)):
            if A[left] > A[left+1]:
                break
        else:
            return A
        right = len(A)-1
        while A[left] <= A[right]:
            right -= 1
        while A[right-1] == A[right]:
            right -= 1
        A[left], A[right] = A[right], A[left]
        return A

File: C++/previous-permutation-with-one-swap.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    vector<int> prevPermOpt1(vector<int>& A) {
        int left = A.size() - 2, right = A.size() - 1;
        for (; left >= 0 && A[left] <= A[left + 1]; --left);
        if (left < 0) {
            return A;
        }
        for (; A[left] <= A[right]; --right);
        for (; A[right - 1] == A[right]; --right);
        swap(A[left], A[right]);
        return A;
    }
};
